README.md

# DIY Streamdeck with the Pimoroni RGB Keypad

This project uses a Raspberry Pi Pico micro controller and an [Pimoroni RGB Keypad](https://shop.pimoroni.com/products/pico-rgb-keypad-base) to provide dynamic app-specific shortcut keys. By monitoring the currently active app on your computer, it automatically loads and displays relevant shortcuts to streamline your workflow and increase productivity.

![Keypad with Zoom Shortcuts](images/keypad.png)

If you find this project helpful please consider giving it a ⭐️ at [GitHub](https://github.com/LennartHennigs/ESPTelnet) and/or [buy me a ☕️](https://ko-fi.com/lennart0815). Thanks!

For the latest changes and the history of changes, please take a look at the [CHANGELOG](https://github.com/LennartHennigs/DIYStreamDeck/blob/main/CHANGELOG.md).

**Note:** This was (and is) a very successful experiment with ChatGPT-4. 🤖 I built this without any knowledge of Python or CircuitPython. The goal was to not program it myself but tell ChatGPT-4 what I wanted. This is the result. It wrote the code and this README as well. This paragraph here is the only piece I am writing myself (and maybe five lines in the CicuitPython code).

## Features

- Allows to assign keyboard shortcuts/key sequences to its keys
- Dynamically detects the active app and updates the keypad with its shortcuts
- Default keypad layout can be defined when no app specific layout is defined
- Alternatively, apps can be launched via the keypad 🆕
- You can also define "folders" - a new keypad definition scheme that can be tied to a single key 🆕
- Keypad definitions are stored and loaded in a JSON configuration file

## Hardware Requirements

- Raspberry Pi Pico
- Pimoroni RGB Keypad for Raspberry Pi Pico
- Micro-USB cable to connect the Pi Pico to your computer

## How it Works

The [`code.py`](https://github.com/LennartHennigs/DIYStreamDeck/blob/main/src/pi_pico/code.py) script reads key definitions from a JSON file and maps them to specific key sequences and LED colors. It listens for the currently active application on the host computer and updates the keypad based on the key mappings for the active application.

The [`watchdog.py`](https://github.com/LennartHennigs/DIYStreamDeck/blob/main/src/mac/watchdog.py) script monitors the currently active application on the host computer and sends its name to the microcontroller connected to the RGB keypad.

## Getting Started

- Install CircuitPython on your Raspberry Pi Pico following the instructions [here](https://learn.adafruit.com/welcome-to-circuitpython/installing-circuitpython).
- Install the required CircuitPython libraries by following the instructions [here](https://learn.adafruit.com/welcome-to-circuitpython/circuitpython-libraries).
- Clone this repository and copy the contents to your Raspberry Pi Pico.
- Edit the [`key_def.json`](https://github.com/LennartHennigs/DIYStreamDeck/blob/main/src/pi_pico/key_def.json) file to configure the shortcut keys and colors for your desired apps.

## Configuration

In the [`key_def.json`](https://github.com/LennartHennigs/DIYStreamDeck/blob/main/src/pi_pico/key_def.json) configuration file, each app is defined as a JSON object with key-value pairs, with two entries: `key_definitions` and `folders`. In the `key_definitions` area the different keys for various apps are defined with the key numbers (0-15). In `folders` key sets can be definied that can be assigned to a key.

These are the possible fields for a key entry:

- `key_sequence`: This field specifies the key combination to be executed when the key is pressed. You can use either a string or an array [to specify the key sequence](https://docs.circuitpython.org/projects/hid/en/latest/_modules/adafruit_hid/keycode.html). If a string is provided, it should contain the keycodes separated by '+' (e.g., "CTRL+ALT+T"). If an array is provided, it should contain the keycodes as separate elements (e.g., ["CTRL", "ALT", "T"]). You can also add delays between key presses within a shortcut by including a floating-point number in the list of keys for a specific shortcut in the key_def.json file. This number represents the delay in seconds between key presses. You can find a list of possible keycodes here.
- `application`: This field is used to specify the application to be launched when an application key is pressed.
- `color`: This field specifies the color of the key, in RGB format. You can specify the color of the key using an RGB string (e.g., "#FF0000" for red, "#00FF00" for green, "#0000FF" for blue).
- `description`: This field provides a description of the function of the key, which is useful for understanding the purpose of each key when printed in the console.
- `folder`: This field allows you assign a "folder" to be opened. The entry also needs the `action` field
- `action`: This field can have the values `open_folder` or `close_folder`. The later is needed inside a folder definition.


With these fields you can define three types of keys, shortcut keys, application launch keys, and folder keys.

- *Shortcut keys* have a `key_sequence` field which specifies the key combination to be executed when the key is pressed.
- *Application keys* have an `application` field which launches the specified application when the key is pressed.
- *Folder keys*  have an `folder` key and an `action: open_folder` field, when the key is pressed it will load the key definitions.

In addition to the regular shortcut keys for an app, there is a special app key called `_otherwise`, which is used as a fallback when no app definition is found in the JSON file. The `_otherwise` key can include its own set of general-purpose shortcut keys, similar to those defined for "App1". When the Python script is running, it constantly monitors the active application on the computer, and if the active application matches any of the keys in the JSON file, the relevant shortcut keys are loaded onto the keypad. If not, the shortcut keys defined under the "_otherwise" key are loaded onto the keypad.

For example, the configuration could look like this:

``` json
{
  "key_definitions": {
    "zoom.us": {
      "0": {
        "key_sequence": [
          "GUI+SHIFT+A"
        ],
        "color": "#FFFF00",
        "description": "Mute/Unmute Audio"
      },
      "1": {
        "key_sequence": [
          "GUI+SHIFT+V"
        ],
        "color": "#FFFF00",
        "description": "Start/Stop Video"
      },
      "15": {
        "key_sequence": [
          "GUI+W",
          0.1,
          "RETURN"
        ],
        "color": "#FF0000",
        "description": "End Meeting"
      }
    },
    "Microsoft Outlook": {
      "0": {
        "key_sequence": [
          "GUI+ONE"
        ],
        "color": "#FFFF00",
        "description": "Switch to Mail"
      },
      "1": {
        "key_sequence": [
          "GUI+TWO"
        ],
        "color": "#FFFF00",
        "description": "Switch to Calendar"
      },
      "3": {
        "key_sequence": [
          "GUI+ALT+N"
        ],
        "color": "#00FF00",
        "description": "New Email"
      }
    },
    "_otherwise": {
      "0": {
        "key_sequence": [
          "GUI+SPACE"
        ],
        "color": "#FFFFFF",
        "description": "Open Spotlight Search"
      },
      "3": {
        "key_sequence": [
          "GUI+LEFT_CONTROL+Q"
        ],
        "color": "#FF0000",
        "description": "Close Application"
      },
      "13": {
        "action": "open_folder",
        "folder": "apps",
        "color": "#FFFFFF",
        "description": "Apps Folder"
      }
    }
  },
  "folders": {
    "apps": {
      "0": {
        "action": "close_folder",
        "color": "#FFFFFF",
        "description": "Close"
      },
      "12": {
        "application": "zoom.us",
        "color": "#0000FF",
        "description": "Launch Zoom"
      },
      "13": {
        "application": "Slack",
        "color": "#FF0000",
        "description": "Launch Slack"
      }
    }
  }
}
```

In the [`key_def.json`](https://github.com/LennartHennigs/DIYStreamDeck/blob/main/src/pi_pico/key_def.json) file, you will find a special app key called `_otherwise`. This key is used to define shortcut keys that are not specific to any particular app. When the Python script is running, it constantly monitors the active application on your computer, and if the active application matches any of the keys in the JSON file, it will load the relevant shortcut keys onto the keypad. If the active application does not match any of the defined keys, the "_otherwise" key is used as a fallback, and the shortcut keys defined under this key are loaded onto the keypad. This means that you can define a set of general-purpose shortcut keys that are always available, regardless of which application is currently active.

## Watchdog

To enable the dynamic detection of the active app, you need to run a watchdog script on your computer that sends the active app's name to the Pi Pico via USB serial. This project includes a Python watchdog script specifically designed for macOS.

To run the watchdog script, navigate to the directory containing the watchdog.py file and execute the following command, e.g.:

``` bash
python3 watchdog.py --port /dev/cu.usbmodem2101 --speed 9600 --verbose
```

- The `--port` parameter needs to be set to the USB serial port corresponding to your Raspberry Pi Pico (e.g., `/dev/cu.usbmodem2101`).
- The optional `--speed` parameter should be set to the desired baud rate for the serial communication (default: `9600`).
- If the optional `--verbose` parameter is set, the current app will be printed to the console.

When the watchdog script detects a change in the active app, it sends the app's name as a single line over the USB serial connection. The Pi Pico then reads this information, loads the corresponding shortcuts from the `key_def.json` file, and updates the keypad accordingly.


CHANGELOG

# CHANGELOG

## Unreleased

- buttons can now launch applications, introduced `application` key to JSON

## Released

### 1.0 – 04-23-2023

- initial version


src/pi_pico/code.py
# DIY Streamdeck code for a Pi Pico - CircuitPython
# L. Hennigs and ChatGPT 4.0
# last changed: 23-05-12
# https://github.com/LennartHennigs/DIYStreamDeck

import time
import json
from rgbkeypad import RgbKeypad
import usb_hid
import usb_cdc
from adafruit_hid.keyboard import Keyboard
from adafruit_hid.keyboard_layout_us import KeyboardLayoutUS
from adafruit_hid.keycode import Keycode
import board


class KeyController:
    # https://docs.circuitpython.org/projects/hid/en/latest/_modules/adafruit_hid/keycode.html
    JSON_FILE = "key_def.json"

    KEYCODE_MAPPING = {name: getattr(Keycode, name) for name in dir(
        Keycode) if not name.startswith("__")}

    def __init__(self):
        self.keypad = RgbKeypad()
        self.keys = self.keypad.keys
        self.keyboard = Keyboard(usb_hid.devices)
        self.layout = KeyboardLayoutUS(self.keyboard)
        self.key_configs, self.folders = self.read_key_configs(self.JSON_FILE)
        self.key_config = self.key_configs.get("_otherwise", {})
        self.usb_serial = usb_cdc.console
        self.update_keys()
        self.folder_open = False
        self.active_app = None
        self.last_key_config = None

    def open_folder(self, folder_name):
        if folder_name in self.folders:
            self.last_key_config = self.key_config
            self.folder_open = True
            self.folder_name = folder_name
            self.key_config = self.folders[folder_name]
            self.update_keys()

    def close_folder(self):
        self.folder_open = False
        self.key_config = self.last_key_config
        self.last_key_config = None
        self.update_keys()

    def key_action(self, key, press=True):
        if key.number in self.key_config:
            key_config_dict = dict(zip(['key_sequences', 'color', 'description',
                                'application', 'action', 'folder'], self.key_config[key.number]))

            key_sequences = key_config_dict['key_sequences']
            color = key_config_dict['color']
            action = key_config_dict.get('action')
            folder = key_config_dict.get('folder')

            self.update_key_led(key, color, press)

            if press:
                if folder:
                    self.open_folder(folder)
                elif action == 'close_folder':
                    self.close_folder()
                else:
                    self.handle_key_sequences(key_sequences, press)
                    if key_config_dict.get('application'):
                        app_name = key_config_dict['application']
                        self.send_application_name(app_name)
            else:
                self.handle_key_sequences(key_sequences, press)

    def update_key_led(self, key, color, press):
        if key.number in self.key_config:
            key_config_dict = dict(zip(['key_sequences', 'color', 'description',
                                   'application', 'action', 'folder'], self.key_config[key.number]))

            color = key_config_dict['color']
            key.set_led(*color) if not press else key.led_off()
            if press:  # Only print the description when the key is pressed
                description = key_config_dict.get('description', '')
                # print(f"Key {key.number} pressed: {description}")

    def handle_key_sequences(self, key_sequences, press):
        for item in key_sequences:
            if isinstance(item, float):
                self.keyboard.release_all()
                time.sleep(item)
            elif isinstance(item, tuple):
                self.keyboard.press(
                    *item) if press else self.keyboard.release(*item)
            else:
                self.keyboard.press(
                    item) if press else self.keyboard.release(item)

    def update_keys(self):
        for key in self.keys:
            if key.number in self.key_config:
                key_config_dict = dict(zip(['key_sequences', 'color', 'description',
                                       'application', 'action', 'folder'], self.key_config[key.number]))

                color = key_config_dict['color']
                key.set_led(*color)
                self.keypad.on_press(key, self.key_action)
                self.keypad.on_release(
                    key, lambda key=key: self.key_action(key, press=False))
            else:
                key.led_off()
                self.keypad.on_press(key, lambda _, key=key: None)
                self.keypad.on_release(key, lambda _, key=key: None)

    def read_serial_line(self):
        if usb_cdc.console.in_waiting > 0:
            raw_data = usb_cdc.console.readline()
            try:
                return raw_data.decode("utf-8").strip()
            except UnicodeDecodeError:
                pass
        return None

    def send_application_name(self, app_name):
        try:
            usb_cdc.console.write(f"Launch: {app_name}\n".encode(
                'utf-8'))  # Encode the string to bytes
            # print(f"Sent to Mac: Launch: {app_name}")
        except Exception as e:
            # print(f"Could not launch {app_name}: {e}\n")
            pass

    def run(self):
        while True:
            raw_app_name = self.read_serial_line()
            if raw_app_name is not None:
                # Split the app_name string on the first occurrence of " ("
                split_app_name = raw_app_name.split(" (", 1)
                # The first part is always the app name
                app_name = split_app_name[0]
                # The second part is the details, if they exist
                url = None
                if len(split_app_name) > 1:
                    # Remove the trailing ")" from the details
                    url = split_app_name[1].rstrip(')')
                # print(f"Active App: {app_name}, URL: {url}")
                self.key_config = self.key_configs.get(
                    app_name, self.key_configs.get("_otherwise", {}))   
                self.update_keys()
            else:
                time.sleep(0.1)
                self.keypad.update()

    def read_key_configs(self, json_filename):
        def convert_keycode_string(keycode_string):
            keycode_list = keycode_string.split('+')
            keycodes = []
            for key in keycode_list:
                if key not in self.KEYCODE_MAPPING:
                    raise ValueError(
                        f"Unknown keycode constant: {key} in '{keycode_string}'")
                keycodes.append(self.KEYCODE_MAPPING[key])
            return tuple(keycodes)

        def convert_color_string(color_string):
            if color_string.startswith("#"):
                return tuple(int(color_string[i:i+2], 16) for i in (1, 3, 5))
            return (0, 0, 0)

        def convert_value(value):
            if isinstance(value, str):
                return convert_keycode_string(value)
            return value

        with open(json_filename, 'r') as json_file:
            json_data = json.load(json_file)

        key_configs = {}
        folders = {}
        for app, configs in json_data["key_definitions"].items():
            key_configs[app] = {}
            for key, config in configs.items():
                key_sequence = config.get('key_sequence', [])
                key_sequences = tuple(convert_value(v) for v in key_sequence) if isinstance(
                    key_sequence, list) else convert_keycode_string(key_sequence)
                color_array = convert_color_string(config.get('color', ''))
                description = config.get('description', '')
                application = config.get('application', '')
                action = config.get('action', '')
                folder = config.get('folder', '')

                if folder and folder not in json_data["folders"]:
                    print(
                        f"Error: Folder '{folder}' not found. Disabling key binding.")
                    key_sequences = ()
                    color_array = (0, 0, 0)
                    description = ''
                    application = ''
                    action = ''
                    folder = ''
                else:
                    key_sequences, color_array, description, application, action, folder = (
                        key_sequences, color_array, description, application, action, folder)

                key_configs[app][int(key)] = (
                    key_sequences, color_array, description, application, action, folder)

        for folder_name, folder_configs in json_data["folders"].items():
            folders[folder_name] = {}
            close_folder_found = False
            for key, config in folder_configs.items():
                key_sequence = config.get('key_sequence', [])
                key_sequences = tuple(convert_value(v) for v in key_sequence) if isinstance(
                    key_sequence, list) else convert_keycode_string(key_sequence)
                color_array = convert_color_string(config.get('color', ''))
                description = config.get('description', '')
                application = config.get('application', '')
                action = config.get('action', '')
                folder = config.get('folder', '')

                if action == "close_folder":
                    close_folder_found = True

                folders[folder_name][int(key)] = (
                    key_sequences, color_array, description, application, action, folder)

            if not close_folder_found:
                raise ValueError(
                    f"Error: Folder '{folder_name}' does not have a 'close_folder' action defined.")

        return key_configs, folders


if __name__ == "__main__":
    controller = KeyController()
    try:
        controller.run()
    except KeyboardInterrupt:
        # turn off all the LEDs when the program is interrupted
        for key in controller.keys:
            key.led_off()
